### PR TITLE
Fix --generate-hashes with bare VCS URLs

### PR DESCRIPTION
**Pull request by jcushman**
#807 introduced an unhandled edge case when bare VCS URLs like `git+git://github.com/jazzband/pip-tools@7d86c8d3ecd1faa6be11c7ddc6b29a30ffd1dae3` are used with `pip-compile --generate-hashes`:

```
$ cat requirements.in
git+https://github.com/jazzband/pip-tools@7d86c8d3ecd1faa6be11c7ddc6b29a30ffd1dae3
$ pip-compile --generate-hashes
Traceback (most recent call last):
  File "/Users/jcwork/.pyenv/versions/pip-tools/bin/pip-compile", line 11, in <module>
    load_entry_point('pip-tools', 'console_scripts', 'pip-compile')()
  File "/Users/jcwork/.pyenv/versions/3.5.3/envs/pip-tools/lib/python3.5/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/Users/jcwork/.pyenv/versions/3.5.3/envs/pip-tools/lib/python3.5/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/Users/jcwork/.pyenv/versions/3.5.3/envs/pip-tools/lib/python3.5/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/jcwork/.pyenv/versions/3.5.3/envs/pip-tools/lib/python3.5/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/Users/jcwork/.pyenv/versions/3.5.3/envs/pip-tools/lib/python3.5/site-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/Users/jcwork/Documents/pip-tools/piptools/scripts/compile.py", line 357, in cli
    hashes = resolver.resolve_hashes(results)
  File "/Users/jcwork/Documents/pip-tools/piptools/resolver.py", line 88, in resolve_hashes
    return {ireq: self.repository.get_hashes(ireq) for ireq in ireqs}
  File "/Users/jcwork/Documents/pip-tools/piptools/resolver.py", line 88, in <dictcomp>
    return {ireq: self.repository.get_hashes(ireq) for ireq in ireqs}
  File "/Users/jcwork/Documents/pip-tools/piptools/repositories/local.py", line 78, in get_hashes
    return self.repository.get_hashes(ireq)
  File "/Users/jcwork/Documents/pip-tools/piptools/repositories/pypi.py", line 299, in get_hashes
    return {self._get_file_hash(cached_link)}
  File "/Users/jcwork/Documents/pip-tools/piptools/repositories/pypi.py", line 323, in _get_file_hash
    with open_local_or_remote_file(location, self.session) as f:
  File "/Users/jcwork/.pyenv/versions/3.5.3/lib/python3.5/contextlib.py", line 59, in __enter__
    return next(self.gen)
  File "/Users/jcwork/Documents/pip-tools/piptools/repositories/pypi.py", line 397, in open_local_or_remote_file
    response = session.get(url, headers=headers, stream=True)
  File "/Users/jcwork/.pyenv/versions/3.5.3/envs/pip-tools/lib/python3.5/site-packages/pip/_vendor/requests/sessions.py", line 546, in get
    return self.request('GET', url, **kwargs)
  File "/Users/jcwork/.pyenv/versions/3.5.3/envs/pip-tools/lib/python3.5/site-packages/pip/_internal/download.py", line 439, in request
    return super(PipSession, self).request(method, url, *args, **kwargs)
  File "/Users/jcwork/.pyenv/versions/3.5.3/envs/pip-tools/lib/python3.5/site-packages/pip/_vendor/requests/sessions.py", line 533, in request
    resp = self.send(prep, **send_kwargs)
  File "/Users/jcwork/.pyenv/versions/3.5.3/envs/pip-tools/lib/python3.5/site-packages/pip/_vendor/requests/sessions.py", line 640, in send
    adapter = self.get_adapter(url=request.url)
  File "/Users/jcwork/.pyenv/versions/3.5.3/envs/pip-tools/lib/python3.5/site-packages/pip/_vendor/requests/sessions.py", line 731, in get_adapter
    raise InvalidSchema("No connection adapters were found for '%s'" % url)
pip._vendor.requests.exceptions.InvalidSchema: No connection adapters were found for 'git+https://github.com/jazzband/pip-tools@7d86c8d3ecd1faa6be11c7ddc6b29a30ffd1dae3'
```

This is because PyPIRepository.get_hashes incorrectly detects all non-editable URLs as hashable, instead of only artifact URLs.

This PR does two things in separate commits (so you can confirm the failing behavior and test case if you want):

* Add a failing test case by having `test_url_package` try all URLs with and without `--generate-hashes`
* Fix the test case by updating get_hashes to use a better hashability check, modeled on the hashability logic in pip's RequirementPreparer.prepare_linked_requirement

**Changelog-friendly one-liner**: Fix --generate-hashes with bare VCS URLs

##### Contributor checklist

- [x] Provided the tests for the changes.
- [x] Requested a review from another contributor.
- [x] Gave a clear one-line description in the PR (that the maintainers can add to CHANGELOG.md on release).
- [x] Assign the PR to an existing or new milestone for the target version (following [Semantic Versioning](https://blog.versioneye.com/2014/01/16/semantic-versioning/)).

### CODE REVIEW

File: piptools/repositories/pypi.py [hunk 1]
```diff
@@ -282,21 +284,29 @@ def get_dependencies(self, ireq):
...
-                cached_link = ireq.link
-            return {self._get_file_hash(cached_link)}
+
+        if ireq.link:
+            link = ireq.link
+
+            if is_vcs_url(link) or (is_file_url(link) and is_dir_url(link)):
+                # Return empty set for unhashable requirements.
+                # Unhashable logic modeled on pip's
+                # RequirementPreparer.prepare_linked_requirement
```
Comment (atugushev) [inline piptools/repositories/pypi.py]: What about this block: https://github.com/pypa/pip/blob/master/src/pip/_internal/operations/prepare.py#L308-L316? Maybe it also should be included here?

Comment (jcushman) [reply piptools/repositories/pypi.py]: That block (`if not req.original_link and not req.is_pinned`) is covered by the logic later in the function. `if is_url_requirement(ireq)` will grab everything with ireq.original_link set, and `if not is_pinned_requirement(ireq)` will handle anything that isn't pinned.

(I compared pip-tools is_pinned_requirement vs. pip's req.is_pinned, and I think the only difference is pip-tools also checks for unpinned requirements like "==1.2.*")

Comment (atugushev) [reply piptools/repositories/pypi.py]: Right!

**Comment by atugushev** [comment]
@jcushman squash please commits.

**Comment by jcushman** [comment]
Squashed!

**Comment by atugushev** [comment]
`pip-tools v3.8.0` is released

**Pull request merged**